src/visitors/derivative_original_visitor.hpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2023 Blue Brain Project, EPFL.
+ * See the top-level LICENSE file for details.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+/**
+ * \file
+ * \brief \copybrief nmodl::visitor::DerivativeOriginalVisitor
+ */
+
+#include "symtab/decl.hpp"
+#include "visitors/ast_visitor.hpp"
+#include <string>
+
+namespace nmodl {
+namespace visitor {
+
+/**
+ * \addtogroup visitor_classes
+ * \{
+ */
+
+/**
+ * \class DerivativeOriginalVisitor
+ * \brief Make a copy of the `DERIVATIVE` block (if it exists), and insert back as
+ * `DERIVATIVE_ORIGINAL_FUNCTION` block.
+ *
+ * If \ref SympySolverVisitor runs successfully, it replaces the original
+ * solution. This block is inserted before that to prevent losing access to
+ * information about the block.
+ */
+class DerivativeOriginalVisitor: public AstVisitor {
+  private:
+    /// The copy of the derivative block we are solving
+    ast::DerivativeBlock* der_block_function = nullptr;
```

Comment:
Why is this a `_function`?

Reply:
Since there's two blocks that will be added: one for `Dx = f(x)` (the `_function` one), and the other one for `Dx = Dx / (1 - dt * J(f))` (the `_jacobian`, will be added in next breakdown of the original PR). I've debated about only adding one block, but the codegen would then require parsing the AST and _modifying_ it, which I feel shouldn't be done at the codegen stage, hence the "two blocks, one visitor" design (the one visitor is here only because it felt pointless to make another one which does almost the same thing).